gax-java/gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java

```diff
@@ -410,6 +439,83 @@ ChannelCredentials createMtlsChannelCredentials() throws IOException, GeneralSec
     return null;
   }
 
+  @VisibleForTesting
+  boolean isGoogleS2AEnabled() {
+    String S2AEnv = envProvider.getenv(S2A_ENV_ENABLE_USE_S2A);
+    boolean isS2AEnv = Boolean.parseBoolean(S2AEnv);
+    if (isS2AEnv) {
+      return true;
+    }
+    return false;
+  }
+
+  @VisibleForTesting
+  boolean shouldUseS2A() {
+    // If EXPERIMENTAL_GOOGLE_API_USE_S2A is not set to true, skip S2A.
+    if (!isGoogleS2AEnabled()) {
+      return false;
+    }
+
+    // If {@link mtlsEndpoint} is not set, skip S2A. S2A is also skipped when there is endpoint
+    // override. Endpoint override is respected when the {@link endpoint} is resolved via AIP#4114,
+    // see EndpointContext.java
+    if (endpoint != mtlsEndpoint) {
+      return false;
```

Comment:
I might be misunderstanding this, but I thought endpoint is the tls endpoint, which is not supposed to be the same as mtlsEndpoint?

e.g if no client certificate is used, then this `endpoint` should be the regular tls endpoint, is that how it works in Java?

Reply:
From what I understand from the logic in [ClientContext.java](https://github.com/googleapis/sdk-platform-java/blob/4d4c798feb6e9eee2c88fdbda1282d8c109b8eb5/gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java#L74), [StubSettings.java](https://github.com/googleapis/sdk-platform-java/blob/4d4c798feb6e9eee2c88fdbda1282d8c109b8eb5/gax-java/gax/src/main/java/com/google/api/gax/rpc/StubSettings.java#L4) and [EndpointContext.java](https://github.com/googleapis/sdk-platform-java/blob/main/gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java), 

The endpoint here is the [resolved endpoint](https://github.com/googleapis/sdk-platform-java/blob/4d4c798feb6e9eee2c88fdbda1282d8c109b8eb5/gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java#L180). Based on [the logic to resolve the endpoint](https://github.com/googleapis/sdk-platform-java/blob/main/gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java#L256), it could be the tls endpoint, or the mtls endpoint. The logic to resolve the endpoint is happening in ClientContext.java, and the resolved endpoint(either tls or mtls) is being passed to the InstantiatingGrpcChannelProvider.

Reply:
Thanks Kui for the offline discussion!
Summary, the logic as it is now is incorrect. We need to only skip S2A if the `mtlsEndpoint` is empty or client override is specified. I'll modify the code to plumb the client override here so it can be inspected.